File: sdks/canary/python/scenarios/d_model_set.py
#!/usr/bin/env python3
# Copyright (C) 2026 Michael Kao
# SPDX-License-Identifier: AGPL-3.0-or-later
"""D-MODEL-SET canary — Python SDK"""

from __future__ import annotations

import json
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), ".."))

from canary import (
    jwt_login,
    Runner,
    Config,
    config_from_env,
    wait_active,
    ensure_session_with_retry,
    raw_do,
)
from llmsafespaces import LLMSafeSpaces


def run(r: Runner, cfg: Config) -> None:
    if not cfg.llm_api_key or not cfg.llm_model:
        r.ok("model-set: skipped (no LLM API key or model)")
        return

    c = LLMSafeSpaces(cfg.api_url, api_key=jwt_login(cfg), timeout=120.0)
    ws_id = None
    cred_id = None
    try:
        ok, ws = r.assert_no_error(
            lambda: c.workspaces.create(
                name="canary-py-model-set", runtime="base", storage_size="1Gi"
            ),
            "create: no error",
        )
        if not ok:
            return
        ws_id = ws.id

        phase = wait_active(c, ws_id)
        r.assert_(phase == "Active", "reach-active", f"got {phase!r}")
        if phase != "Active":
            return

        cred_value = json.dumps(
            {
                "kind": cfg.llm_provider,
                "slug": "canary-py-model-set",
                "apiKey": cfg.llm_api_key,
            }
        )
        ok_cred, cred = r.assert_no_error(
            lambda: c.secrets.create(
                name="canary-py-model-set-cred", type="llm-provider", value=cred_value
            ),
            "create-cred: no error",
        )
        if ok_cred and cred:
            cred_id = cred.id
            r.assert_no_error(
                lambda: c.workspaces.set_bindings(ws_id, [cred_id]),
                "bind-cred: no error",
            )

        r.assert_no_error(
            lambda: c.workspaces.set_model(ws_id, cfg.llm_model),
            "set-model: no error",
        )

        try:
            sess = ensure_session_with_retry(c, ws_id, 5)
        except Exception as e:
            r.fail("ensure-session: no error", str(e))
            return
        r.ok("ensure-session: no error")
        sid = sess.sessionId

        ok2, msg = r.assert_no_error(
            lambda: c.sessions.send_message(ws_id, sid, "Reply with exactly: MODEL-OK"),
            "send-message: no error",
        )
        if ok2 and msg is not None:
            r.assert_(len(msg.content) > 0, "send-message: non-empty content")
            r.assert_(
                "MODEL-OK" in msg.content.upper(),
                "send-message: contains MODEL-OK",
                repr(msg.content[:100]),
            )

        # Negative: empty model
        s1, _ = raw_do(
            "PUT",
            f"{cfg.api_url}/api/v1/workspaces/{ws_id}/model",
            cfg.api_key,
            json.dumps({"model": ""}).encode(),
        )
        r.assert_(s1 in (400, 422), "empty-model: 400 or 422", str(s1))

        # Negative: nonexistent workspace
        s2, _ = raw_do(
            "PUT",
            f"{cfg.api_url}/api/v1/workspaces/00000000-0000-0000-0000-000000000000/model",
            cfg.api_key,
            json.dumps({"model": "test/model"}).encode(),
        )
        r.assert_(s2 == 404, "nonexistent-ws-model: 404", str(s2))

        # Negative: bad model — verify ws still Active
        ok3, ws_check = r.assert_no_error(
            lambda: c.workspaces.get(ws_id), "ws-still-exists: no error"
        )
        if ok3 and ws_check:
            r.assert_(
                ws_check.phase == "Active",
                "after-bad-model: ws still Active",
                ws_check.phase,
            )

    finally:
        if cred_id:
            try:
                c.secrets.delete(cred_id)
            except Exception:
                pass
        if ws_id:
            try:
                c.workspaces.delete(ws_id)
            except Exception:
                pass


if __name__ == "__main__":
    r = Runner("model-set")
    cfg = config_from_env()
    run(r, cfg)
    r.print()
    sys.exit(r.exit_code())

File: sdks/canary/typescript/scenarios/d-model-set.ts
// Copyright (C) 2026 Michael Kao
// SPDX-License-Identifier: AGPL-3.0-or-later
// D-MODEL-SET canary — TypeScript SDK

import { LLMSafeSpaces } from '../../../typescript/src/index.js';
import { Runner, Config, configFromEnv, nodeFetch, waitActive, ensureSessionWithRetry } from '../canary.js';

async function run(r: Runner, cfg: Config): Promise<void> {
  if (!cfg.llmApiKey || !cfg.llmModel) {
    r.ok('model-set: skipped (no LLM API key or model)');
    return;
  }
  const c = new LLMSafeSpaces({ baseUrl: cfg.apiUrl, apiKey: cfg.apiKey, timeout: 120000, fetch: nodeFetch as any });
  let wsId: string | null = null;
  try {
    const [ok, ws] = await r.assertNoError(
      () => c.workspaces.create({ name: 'canary-ts-model-set', runtime: 'base', storageSize: '1Gi' }),
      'create: no error');
    if (!ok || !ws) return;
    wsId = ws.id;

    const phase = await waitActive(c, wsId);
    r.assert(phase === 'Active', 'reach-active', `got "${phase}"`);
    if (phase !== 'Active') return;

    await r.assertNoError(
      () => c.workspaces.setModel(wsId!, cfg.llmModel), 'set-model: no error');

    const [ok2, models] = await r.assertNoError(
      () => c.workspaces.getModels(wsId!), 'get-models-after-set: no error');
    if (ok2 && models) {
      r.assert(models.currentModel === cfg.llmModel, 'models: currentModel updated',
        `${models.currentModel} vs ${cfg.llmModel}`);
      const selected = (models.models as any[])?.find((m: any) => m.selected);
      r.assert(selected?.id === cfg.llmModel, 'models: selected matches',
        `${selected?.id} vs ${cfg.llmModel}`);
    }

    const sess = await ensureSessionWithRetry(c, wsId!, 5);
    r.ok('ensure-session: no error');
    const [ok3, msg] = await r.assertNoError(
      () => c.sessions.sendMessage(wsId!, sess.sessionId, 'Reply with exactly: MODEL-SET-OK'),
      'send-message: no error');
    if (ok3 && msg) {
      r.assert((msg.text ?? (msg.parts ?? []).filter(p => p.type === 'text').map(p => p.text ?? '').join('')).length > 0, 'send-message: non-empty content');
    }

    await r.assertError(
      () => c.workspaces.setModel(wsId!, ''), 'set-empty-model: error');

    await r.assertError(
      () => c.workspaces.setModel('00000000-0000-0000-0000-000000000000', cfg.llmModel),
      'set-model-nonexistent-ws: error');

    const [ok4] = await r.assertNoError(
      () => c.workspaces.setModel(wsId!, cfg.badModel).catch(async (e: any) => {
        if (e?.status === 400 || e?.message) {
          const ws = await c.workspaces.get(wsId!);
          r.assert(ws.phase === 'Active', 'bad-model: ws still Active after bad model',
            ws.phase);
          throw e;
        }
        throw e;
      }),
      'set-bad-model: handled');
    if (!ok4) {
      const ws2 = await c.workspaces.get(wsId!);
      r.assert(ws2.phase === 'Active', 'bad-model-fallback: ws still Active', ws2.phase);
    }

  } finally {
    if (wsId) { try { await c.workspaces.delete(wsId); } catch {} }
  }
}

async function main() {
  const r = new Runner('model-set');
  await run(r, configFromEnv());
  r.print();
  process.exit(r.exitCode());
}
main().catch(e => { console.error(e); process.exit(1); });
